From bb46e0e2c1bb2bda44f976e9166e89ad37e6cb46 Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@linaro.org>
Date: Wed, 6 Apr 2016 15:57:29 +0200
Subject: [PATCH 10/13] STM: iec61937: force hw_param channels to stereo

In case of pass-through, iec frames are sent over a stereo PCM
stream. So alsa device should be configured with channels = 2.
This is needed to be compatible with SPDIF electrical/optical output
that supports only stereo in PCM mode and up to 8 channels in iec61937 mode.

Change-Id: I23e5cfb21b3a63dbb3b4d48c2639406adae28a1f
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>
---
 ext/alsa/gstalsasink.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/ext/alsa/gstalsasink.c b/ext/alsa/gstalsasink.c
index a5e8605..822da94 100644
--- a/ext/alsa/gstalsasink.c
+++ b/ext/alsa/gstalsasink.c
@@ -810,12 +810,15 @@ alsasink_parse_spec (GstAlsaSink * alsa, GstAudioRingBufferSpec * spec)
         default:
           goto error;
       }
+      alsa->channels = GST_AUDIO_INFO_CHANNELS (&spec->info);
       break;
     case GST_AUDIO_RING_BUFFER_FORMAT_TYPE_A_LAW:
       alsa->format = SND_PCM_FORMAT_A_LAW;
+      alsa->channels = GST_AUDIO_INFO_CHANNELS (&spec->info);
       break;
     case GST_AUDIO_RING_BUFFER_FORMAT_TYPE_MU_LAW:
       alsa->format = SND_PCM_FORMAT_MU_LAW;
+      alsa->channels = GST_AUDIO_INFO_CHANNELS (&spec->info);
       break;
     case GST_AUDIO_RING_BUFFER_FORMAT_TYPE_AC3:
     case GST_AUDIO_RING_BUFFER_FORMAT_TYPE_EAC3:
@@ -823,13 +826,14 @@ alsasink_parse_spec (GstAlsaSink * alsa, GstAudioRingBufferSpec * spec)
     case GST_AUDIO_RING_BUFFER_FORMAT_TYPE_MPEG:
       alsa->format = SND_PCM_FORMAT_S16_BE;
       alsa->iec958 = TRUE;
+      /* Set channels to stereo as iec playload are rendered on a stereo stream */
+      alsa->channels = 2;
       break;
     default:
       goto error;
 
   }
   alsa->rate = GST_AUDIO_INFO_RATE (&spec->info);
-  alsa->channels = GST_AUDIO_INFO_CHANNELS (&spec->info);
   alsa->buffer_time = spec->buffer_time;
   alsa->period_time = spec->latency_time;
   alsa->access = SND_PCM_ACCESS_RW_INTERLEAVED;
-- 
2.7.4

